### DISCUSSION TITLE
hyperlinks to slides

### DISCUSSION THREAD

**Discussion by obsaditelnost** [davidgohel/flextable#424] [Q&A] [answered]
I try to build some kind of TOC in Powerpoint with the phantastic offier package. Working with placeholders and ph_slidelink might be possible but laborious. ftext also does not have the possibility to support links. 

In flextable there seems to be a function hyperlink_text() - but only for Word as far as I can see. Is is possible to use it for slide links or is there any other way to achieve this task?

```
library(officer)
library(flextable)
 
# hyperlink_text() for pptx slides? 
df <- flextable(data.frame(topic = c("topic 1", "topic 2"), 
                 begins_at_slide = c(1,3)))

loc_title <- ph_location_type(type = "title")
loc_body <- ph_location_type(type = "body")
doc <- read_pptx()
doc <- add_slide(doc)
doc <- ph_with(x = doc, "topic 1", location = loc_title)
doc <- add_slide(doc)
doc <- ph_with(x = doc, "topic 1", location = loc_title)
doc <- add_slide(doc)
doc <- ph_with(x = doc, "topic 2", location = loc_title)
doc <- add_slide(doc, layout = "Title and Content", master = "Office Theme")
doc <- ph_with(x = doc, "TOC", location = loc_title)
doc <- ph_with(x = doc, df, location = loc_body)
print(doc, target = "test.pptx" )
```

**Comment by davidgohel** [comment]
hello

Few comments/questions:

* yes, `ftext()` does not let you add links, but as indicated in the manuals `hyperlink_ftext()` does let you add hyperlinks.
* why do you say "function hyperlink_text() - but only for Word as far as I can see"? The manual does not state it is only for Word, I am worry why users would think it is only for Word. IMO, nothing says so in the doc. I would like to remove the ambiguity on this point.


What you are trying to do seems very hard to me, unless we implement a derivate of `ph_slidelink()` for flextable, which would be very hard and only for PowerPoint output. 

Is PowerPoint the only option you have? Why not building an HTML version of your presentation. HTML is a very good format and usually provides more options than pptx (in that case, `hyperlink_ftext()` could let you do the job).

Your remark about `ph_slidelink()` seems hard to me;) This is not that hard, I find it easier to use it than to develop a new feature just for enabling building TOC that is not a PowerPoint feature. All solutions can only be workarounds, I recommend in general to stick to what exists and not try to invent something that is not possible natively.

This is an example using `ph_slidelink()`. It's available now, it does not need new development, it is *PowerPoint native*.


```r

loc_title <- ph_location_type(type = "title")

doc <- read_pptx()

doc <- add_slide(doc)
doc <- ph_with(x = doc, "youp", location = ph_location(top = 0, newlabel = "a"))
doc <- ph_with(x = doc, "yoip", location = ph_location(top = 1, newlabel = "b"))
doc <- ph_with(x = doc, "yois", location = ph_location(top = 2, newlabel = "c"))

doc <- add_slide(doc)
doc <- ph_with(x = doc, "dummy title 1", location = loc_title)
doc <- add_slide(doc)
doc <- ph_with(x = doc, "dummy title 2", location = loc_title)
doc <- add_slide(doc)
doc <- ph_with(x = doc, "dummy title 3", location = loc_title)

doc <- on_slide(doc, 1)
doc <- ph_slidelink(x = doc, ph_label = "a", slide_index = 2)
doc <- ph_slidelink(x = doc, ph_label = "b", slide_index = 3)
doc <- ph_slidelink(x = doc, ph_label = "c", slide_index = 4)

print(doc, target = "fileout.pptx" )
```


What do you think of this?

**Comment by obsaditelnost** [comment, marked as answer]
Thanks David. Sometimes you have to use PPTX because "it's the way we do things, just do it". 

With respect to hyperlink_text() the [Note in the reference ](https://davidgohel.github.io/flextable/reference/hyperlink_text.html) says something about requiring word as output format (maybe I just misunderstood the intention of the note). So i didn't try to use it in Powerpoint. Now i tried, and: it works!

Changing the package code to support internal links would mean some struggle with the "action" attribute of the a:hlinkClick XML-tag, but given the hyperlink_text()-functionality, I can already use the "#"-feature as a workaround which works with MS products (I guess it's not standard compliant though)

```
library(officer)
library(flextable)

df <- flextable(data.frame(topic = c("topic 1", "topic 2"), 
                           begins_at_slide = c(1,3)))
df <- compose(x = df, 
              j = "begins_at_slide",
              value = as_paragraph(
                hyperlink_text(x=begins_at_slide, 
                               url = paste0('#', begins_at_slide) ) ) )

loc_title <- ph_location_type(type = "title")
loc_body <- ph_location_type(type = "body")
doc <- read_pptx()
doc <- add_slide(doc)
doc <- ph_with(x = doc, "topic 1", location = loc_title)
doc <- add_slide(doc)
doc <- ph_with(x = doc, "topic 1", location = loc_title)
doc <- add_slide(doc)
doc <- ph_with(x = doc, "topic 2", location = loc_title)
doc <- add_slide(doc, layout = "Title and Content", master = "Office Theme")
doc <- ph_with(x = doc, "TOC", location = loc_title)
doc <- ph_with(x = doc, df, location = loc_body)
print(doc, target = "test.pptx" )
```

**Comment by davidgohel** [comment]
> Sometimes you have to use PPTX because "it's the way we do things, just do it".

yeah, I am ok with that. 

This is great, I didn't know using `#` was a valid option when output is PowerPoint! 

> I guess it's not standard compliant though

I really don't know. I am almost sure I did not do anything about that. I am surprised it works, but I can reproduce and see it works. :)

Also thanks for the feedback about the note. I will rephrase it. When I wrote that, I meant that if working with "R Mardown" and a Word output, the chunk would not be rendered correctly if `officedown::rdocx_document()` is not used.

**Reply by davidgohel** [reply]
I think I can mark your solution as "answer". Are you ok?
